### PR TITLE
SP-6208: granfa get panel step

### PR DESCRIPTION
**Pull request by itaischwartz**

### CODE REVIEW

File: steps/grafana/get-panel/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    required: true
+    example: "https://play.grafana.org/render/d-solo/000000012/grafana-play-home?orgId=1&from=1620203934175&to=1620207534175&panelId=2&width=1000&height=500&tz=Asia%2FJerusalem"
+  - name: API_KEY
+    description: "API key for a particular user (leave unset if unauthenticated)"
+    example: "eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk"
+
+outputs:
+  - name: output
+    description: "base64 rendered image"
+    type: "string"
```
Comment (kostyay) [inline steps/grafana/get-panel/manifest.yaml:21]: this is not correct as your output goes to root

**Review by kostyay — APPROVED** [review]
fix the output and merge

File: steps/grafana/get-panel/main.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	output, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return step.ExitCodeFailure, nil, fmt.Errorf("readPanel: %w", err)
+	}
+
+	if err = upload.RichOutput(ctx, ioutil.NopCloser(bytes.NewBuffer(output)), upload.ContentTypePNG); err != nil {
+		return step.ExitCodeFailure, nil, fmt.Errorf("uploadPanel: %w", err)
+	}
+
+	result := []byte(base64.StdEncoding.EncodeToString(output))
```
Comment (hagaishapira) [inline steps/grafana/get-panel/main.go:75]: You are not using the outputs struct you've defined

File: steps/grafana/get-panel/main.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+	if l.args.API_KEY != "" {
+		// add authorization header with api key to the req
+		var bearer = "Bearer " + l.args.API_KEY
+		request.Header.Add("Authorization", bearer)
+	}
+
+	resp, err := client.Do(request)
+	if err != nil {
+		return step.ExitCodeFailure, nil, fmt.Errorf("get: %w", err)
```
Comment (hagaishapira) [inline steps/grafana/get-panel/main.go:62]: What about a non 2XX response?

File: steps/grafana/get-panel/main.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	args Args
+}
+
+type Outputs struct {
+	step.Outputs
+	URL string `json:"url"`
+}
+
+func (l *GrafanaGetPanel) Init() error {
+	err := envconf.Parse(&l.args)
```
Comment (hagaishapira) [inline steps/grafana/get-panel/main.go:36]: Use https://github.com/stackpulse/steps-sdk-go/blob/master/env/parse Parse and not github.com/caarlos0/env/v6 directly

File: steps/grafana/get-panel/manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  type: remediation
+  version: 1.0.0
+envs:
+  - name: URL
+    description: "Grafana Direct link rendered image URL"
+    required: true
+    example: "https://play.grafana.org/render/d-solo/000000012/grafana-play-home?orgId=1&from=1620203934175&to=1620207534175&panelId=2&width=1000&height=500&tz=Asia%2FJerusalem"
+  - name: API_KEY
+    description: "API key for a particular user (leave unset if unauthenticated)"
+    example: "eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk"
```
Comment (hagaishapira) [inline steps/grafana/get-panel/manifest.yaml:16]: This is not a valid key to the grafana url you've provided, right?

File: steps/grafana/get-panel/manifest.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+outputs:
+  - name: output
```
Comment (hagaishapira) [inline steps/grafana/get-panel/manifest.yaml:19]: I prefer the field name to be more explicit and self explaining like "encoded_image"

File: steps/grafana/get-panel/manifest.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+apiVersion: stackpulse.io/v1
+kind: Step
+metadata:
+  name: "grafana_get_panel"
+  description: "Get a panel from grafana as PNG."
```
Comment (hagaishapira) [inline steps/grafana/get-panel/manifest.yaml:5]: Perhaps expend on the fact that this uploads the panel to StackPulse's file storage and encodes the image as base64

**Review by hagaishapira — CHANGES_REQUESTED** [review]

File: steps/grafana/get-panel/main.go (hunk 4 above)
Comment (itaischwartz) [reply steps/grafana/get-panel/main.go:36]: done

File: steps/grafana/get-panel/main.go (hunk 3 above)
Comment (itaischwartz) [reply steps/grafana/get-panel/main.go:62]: done

File: steps/grafana/get-panel/main.go (hunk 2 above)
Comment (itaischwartz) [reply steps/grafana/get-panel/main.go:75]: done

File: steps/grafana/get-panel/manifest.yaml (hunk 5 above)
Comment (itaischwartz) [reply steps/grafana/get-panel/manifest.yaml:16]: random key

File: steps/grafana/get-panel/manifest.yaml (hunk 6 above)
Comment (itaischwartz) [reply steps/grafana/get-panel/manifest.yaml:19]: done

File: steps/grafana/get-panel/manifest.yaml (hunk 1 above)
Comment (itaischwartz) [reply steps/grafana/get-panel/manifest.yaml:21]: done

File: steps/grafana/get-panel/manifest.yaml (hunk 7 above)
Comment (hagaishapira) [inline steps/grafana/get-panel/manifest.yaml:5]:
```suggestion
  description: "Get a panel from grafana using a Grafana Image Renderer url. The image will be uploaded to StackPulse's file storage and outputted from the step encoded to base64."
```

File: steps/grafana/get-panel/main.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+type Args struct {
+	URL     string `env:"URL,required"`
+	API_KEY string `env:"API_KEY", envDefault:""`
+}
+
+type GrafanaGetPanel struct {
+	args Args
+}
+
+type Outputs struct {
```
Comment (hagaishapira) [inline steps/grafana/get-panel/main.go]: delete this as you're not using it

**Review by hagaishapira — APPROVED** [review]
Fix texts and last comment and MWG

**Pull request merged**